src/stdlib/from_unix_timestamp.rs

```diff
@@ -0,0 +1,269 @@
+                source: "from_unix_timestamp!([])",
+                result: Err(
+                    r#"function call error for "from_unix_timestamp" at (0:17): unable to coerce array into timestamp"#,
+                ),
+            },
+            Example {
+                title: "object",
+                source: "from_unix_timestamp!({})",
+                result: Err(
+                    r#"function call error for "from_unix_timestamp" at (0:17): unable to coerce object into timestamp"#,
+                ),
+            },
+            Example {
+                title: "regex",
+                source: "from_unix_timestamp!(r'foo')",
+                result: Err(
+                    r#"function call error for "from_unix_timestamp" at (0:21): unable to coerce regex into timestamp"#,
+                ),
+            },
+        ]
+    }
+
+    fn compile(
+        &self,
+        _state: &state::TypeState,
+        _ctx: &mut FunctionCompileContext,
+        arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+
+        let unit = arguments
+            .optional_enum("unit", Unit::all_value().as_slice())?
+            .map(|s| {
+                Unit::from_str(&s.try_bytes_utf8_lossy().expect("unit not bytes"))
+                    .expect("validated enum")
+            })
+            .unwrap_or_default();
+
+        Ok(FromUnixTimestampFn { value, unit }.as_expr())
+    }
+}
+
+#[derive(Default, Debug, Clone, Copy, PartialEq, Eq)]
+enum Unit {
+    #[default]
+    Seconds,
+    Milliseconds,
+    Nanoseconds,
+}
+
+impl Unit {
+    fn all_value() -> Vec<Value> {
+        use Unit::{Milliseconds, Nanoseconds, Seconds};
+
+        vec![Seconds, Milliseconds, Nanoseconds]
+            .into_iter()
+            .map(|u| u.as_str().into())
+            .collect::<Vec<_>>()
+    }
+
+    const fn as_str(self) -> &'static str {
+        use Unit::{Milliseconds, Nanoseconds, Seconds};
+
+        match self {
+            Seconds => "seconds",
+            Milliseconds => "milliseconds",
+            Nanoseconds => "nanoseconds",
+        }
+    }
+}
+
+impl FromStr for Unit {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        use Unit::{Milliseconds, Nanoseconds, Seconds};
+
+        match s {
+            "seconds" => Ok(Seconds),
+            "milliseconds" => Ok(Milliseconds),
+            "nanoseconds" => Ok(Nanoseconds),
+            _ => Err("unit not recognized"),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct FromUnixTimestampFn {
+    value: Box<dyn Expression>,
+    unit: Unit,
+}
+
+impl FunctionExpression for FromUnixTimestampFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let unit = self.unit;
+        from_unix_timestamp(value, unit)
+    }
+
+    fn type_def(&self, state: &state::TypeState) -> TypeDef {
+        self.value
```

Comment:
This is not quite right.
- `.fallible_unless(Kind::timestamp())` makes the type definition fallible unless the input kind is a timestamp. There are a few issues with this. First, the VRL compiler automatically checks function argument types, so you don't need to worry about checking those here. Second, `timestamp` is an invalid type here anyway. 
- `with_kind` completely overwrites the original kind (`self.value`) here, so there's no need to use `self.value` at all

There's 2 parts to a type definition, so lets look at the individually:
- `kind`: This is the actual type returned by the function. This function will always return a timestamp, so it can just be hard-coded to `Kind::timestamp()`.
- `fallibility`: This is a boolean, saying if the function can fail (at runtime) or not. There's an interesting edge case here that I'll talk about below, but in general this function can always potentially fail. If you want to just hard-code this to fallible I'd be okay with that, so the entire type def becomes `TypeDef::timestamp().fallible()`

If we can prove at (VRL) compile time that a function will not fail, it's nice to change the type definition in that case so the user doesn't have to try to handle errors that can't happen. This usually applies to arguments where the most common pattern is to pass in a literal, such as the "unit" argument. However, even if we did that, the timestamp conversion itself can usually fail. There is a single edge case where it is known that `nanoseconds` is hard-coded as the unit, that this function is infallible. If you want to try to handle that case, I can give you an example to follow, however that seems like it's a small enough edge case that it might not really be helpful to users and might just end up confusing them more.